openrr-remote/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+            &request.to,
+            request.time.unwrap().try_into().unwrap(),
+        )
+        .map_err(|e| tonic::Status::unknown(e.to_string()))?
+        .into();
+        Ok(tonic::Response::new(p))
+    }
+}
+
+// =============================================================================
+// arci::Gamepad
+
+remote_types!(Gamepad {
+    client: RemoteGamepadSender,
+    server: RemoteGamepadReceiver,
+});
+
+#[arci::async_trait]
+impl arci::Gamepad for RemoteGamepadSender {
+    async fn next_event(&self) -> arci::gamepad::GamepadEvent {
+        let mut client = self.client.clone();
+        match client.next_event(()).await {
+            Ok(event) => event.into_inner().into(),
+            Err(e) => {
+                error!("{}", e);
+                arci::gamepad::GamepadEvent::Unknown
+            }
+        }
+    }
+
+    fn stop(&self) {
+        let mut client = self.client.clone();
+        if let Err(e) = block_in_place(client.stop(())) {
+            error!("{}", e);
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<C> gamepad_server::Gamepad for RemoteGamepadReceiver<C>
+where
+    C: arci::Gamepad + 'static,
+{
+    async fn next_event(
+        &self,
+        _: tonic::Request<()>,
+    ) -> Result<tonic::Response<pb::GamepadEvent>, tonic::Status> {
+        Ok(tonic::Response::new(self.inner.next_event().await.into()))
+    }
+
+    async fn stop(&self, _: tonic::Request<()>) -> Result<tonic::Response<()>, tonic::Status> {
+        self.inner.stop();
+        Ok(tonic::Response::new(()))
+    }
+}
+
+// =============================================================================
+// Messages
+
+impl From<arci::TrajectoryPoint> for pb::TrajectoryPoint {
+    fn from(val: arci::TrajectoryPoint) -> Self {
+        Self {
+            positions: val.positions,
+            velocities: val.velocities.unwrap_or_default(),
+            time_from_start: Some(val.time_from_start.into()),
+        }
+    }
+}
+
+impl From<pb::TrajectoryPoint> for arci::TrajectoryPoint {
+    fn from(val: pb::TrajectoryPoint) -> Self {
+        Self {
+            positions: val.positions,
+            velocities: if val.velocities.is_empty() {
+                None
+            } else {
+                Some(val.velocities)
+            },
+            time_from_start: val.time_from_start.unwrap().try_into().unwrap(),
+        }
+    }
+}
+
+impl From<arci::BaseVelocity> for pb::BaseVelocity {
+    fn from(val: arci::BaseVelocity) -> Self {
+        Self {
+            x: val.x,
+            y: val.y,
+            theta: val.theta,
+        }
+    }
+}
+
+impl From<pb::BaseVelocity> for arci::BaseVelocity {
+    fn from(val: pb::BaseVelocity) -> Self {
+        Self {
+            x: val.x,
+            y: val.y,
+            theta: val.theta,
+        }
+    }
+}
+
+impl From<arci::Isometry2<f64>> for pb::Isometry2 {
+    fn from(val: arci::Isometry2<f64>) -> Self {
+        Self {
+            rotation: Some(pb::UnitComplex {
```

Comment:
why we need Option?

Reply:
In proto3, all fields are optional: https://github.com/protocolbuffers/protobuf/blob/master/docs/field_presence.md

Reply:
I see. I know that.